packages/languages/src/node/language-server-contribution.ts

```diff
@@ -27,18 +28,34 @@ export {
 export const LanguageServerContribution = Symbol('LanguageServerContribution');
 export interface LanguageServerContribution extends LanguageContribution {
     start(clientConnection: IConnection): void;
+    /**
+     * Called on shut down. Only synchronous operations are permitted.
+     */
+    stop?(): void;
 }
 
 @injectable()
 export abstract class BaseLanguageServerContribution implements LanguageServerContribution {
 
     abstract readonly id: string;
     abstract readonly name: string;
-    abstract start(clientConnection: IConnection): void;
+    protected readonly spawnedProcessIds: number[];
 
     @inject(RawProcessFactory)
     protected readonly processFactory: RawProcessFactory;
 
+    @inject(ProcessManager)
+    protected readonly processManager: ProcessManager;
+
+    abstract start(clientConnection: IConnection): void;
+
+    stop(): void {
+        this.spawnedProcessIds
+            .map(id => this.processManager.get(id))
+            .filter(process => process)
+            .forEach(process => this.processManager.delete(process!));
```

Comment:
Should we wait for the processes to be killed before exiting ?

We can't be sure that just sending SIGTERM here will kill them...

Reply:
Yes, each call must be synchronous. https://nodejs.org/api/process.html#process_event_exit

Reply:
I mean the call to kill I think is synchronous... but the signal handler of the processes is not...
So you send kill to all .. the main process exists.. but the childs will live on until they've handled the SIGTERM.

And if they don't exit properly with SIGTERM you will have lost them.
So maybe we should wait for their exit event before ending this operation.

Reply:
LS processes should exit automatically when the parent process is not alive, look for InitializeParams.processId comment in the lsp. If it does not happen it means that the worker process is alive. One should check it and if it is the case what the forked worker does not exit when the master exits, then changes in this pr won’t help since the worker is still running and the proper fix will be to explicitly stop running workers on exit in the master process.

Reply:
> LS processes should exit automatically when the parent process is not alive

That is not happening: https://github.com/theia-ide/theia/issues/815#issuecomment-343487901

> the worker process is alive

I cannot see other Theia processes.

Reply:
I am happy to share how to set up things on Windows if someone would like to reproduce the defect.

Reply:
Ok, then it should help 👍🏻Java ls is bogus

Reply:
You mean, in general, any LS started from a jar?

Reply:
Or we use outdated java ls, I can see in their code handling of this case: https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/LanguageServer.java

Reply:
But the screenshot (gif whatever) I have attached to this thread does not do anything with the Java LS. It starts the Yang LS.

Reply:
I see, I thought lsp4j or Xtext ls supports it, but cannot find such code there. Btw the lsp says that shutdown + exit requests should be sent to shutdown the server.

Reply:
https://github.com/eclipse/xtext-core/issues/552 - a bug for it in Xtext